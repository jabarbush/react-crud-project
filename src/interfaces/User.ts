interface User {
    id: number;
    name: string;
    dob: string;
    phone: string;
    email: string;
    signature?: {
        pin: string;
        fontStyle: string;
    } | null;
  }
  
  export default User;
  