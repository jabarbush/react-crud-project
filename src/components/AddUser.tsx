import React, { useState } from 'react';
import userData from '../db.json';
import User from '../interfaces/User';
import AddUserProps from '../interfaces/AddUserProps';

const AddUser: React.FC<AddUserProps> = ({ onClose, onAddUser }) => {
  const [isPopupVisible, setIsPopupVisible] = useState(true);
  const [name, setName] = useState('');
  const [dob, setDob] = useState('');
  const [phone, setPhone] = useState('');
  const [email, setEmail] = useState('');

  const handleClosePopup = () => {
    setIsPopupVisible(false);
    onClose();
  };

  if (!isPopupVisible) {
    return null;
  }

  const handleAddUser = (e: React.FormEvent) => {
    e.preventDefault();

    if (!name || !dob || !phone || !email) {
      alert('Please fill in all required fields.');
      return;
    }

    const phoneRegex = /^\d{10}$/;
    if (!phone.match(phoneRegex)) {
      alert('Please enter a valid phone number.');
      return;
    }

    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!email.match(emailRegex)) {
      alert('Please enter a valid email address.');
      return;
    }

    const dateRegex = /^(0[1-9]|1[0-2])\/(0[1-9]|1\d|2\d|3[01])\/(19|20)\d{2}$/;
    if (!dob.match(dateRegex)) {
      alert('Please enter a valid date in the format MM/DD/YYYY.');
      return;
    }

    const newUser: User = {
      id: Math.max(...userData.users.map((user) => user.id)) + 1,
      name,
      dob,
      phone,
      email,
      signature: null
    };

    onAddUser(newUser);
  };

  return (
    <div className="add-user-popup">
      <form onSubmit={handleAddUser}>
        <h2 className="new-user-title">Add New User</h2>
        <label>Name</label>
        <input type="text" placeholder="Name" value={name} onChange={(e) => setName(e.target.value)} required />
        <label>Date of Birth</label>
        <input type="text" placeholder="Date of Birth" value={dob} onChange={(e) => setDob(e.target.value)} required />
        <label>Phone</label>
        <input type="text" placeholder="Phone" value={phone} onChange={(e) => setPhone(e.target.value)} required />
        <label>Email</label>
        <input type="text" placeholder="Email" value={email} onChange={(e) => setEmail(e.target.value)} required />
        <button className="cancel" onClick={handleClosePopup}>
          Cancel
        </button>
        <button className="add-user-submit" type="submit">
          Add User
        </button>
      </form>
    </div>
  );
};

export default AddUser;
