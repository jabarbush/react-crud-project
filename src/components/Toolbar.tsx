import React, { useEffect, useState } from "react";
import AddUser from "./AddUser";
import User from "../interfaces/User";
import ToolbarProps from "../interfaces/ToolbarProps";

const Toolbar: React.FC<ToolbarProps> = ({ userList, setUserList }) => {
  const [isPopupOpen, setIsPopupOpen] = useState(false);
  const [isSortOpen, setIsSortOpen] = useState(false);
  const [isFilterOpen, setIsFilterOpen] = useState(false);
  const [sortOrder, setSortOrder] = useState("asc");

  useEffect(() => {
    fetch("http://localhost:3001/users")
      .then((response) => response.json())
      .then((data) => setUserList(data))
      .catch((error) => console.error(error));
  }, [setUserList]);

  const handleOpenPopup = () => {
    setIsPopupOpen(true);
  };

  const handleClosePopup = () => {
    setIsPopupOpen(false);
  };

  const handleAddUser = (user: User) => {
    fetch("http://localhost:3001/users", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(user),
    })
      .then((response) => response.json())
      .then((data) => {
        setUserList((prevList) => [...prevList, data]);
        setIsPopupOpen(false);
      })
      .catch((error) => console.error(error));
    setIsPopupOpen(false);
  };

  const handleSortAscending = () => {
    const sortedList = [...userList].sort((a, b) =>
      new Date(a.dob).getTime() - new Date(b.dob).getTime()
    );
    setUserList(sortedList);
    setIsSortOpen(false);
    setSortOrder("asc");
  };

  const handleSortDescending = () => {
    const sortedList = [...userList].sort((a, b) =>
      new Date(b.dob).getTime() - new Date(a.dob).getTime()
    );
    setUserList(sortedList);
    setIsSortOpen(false);
    setSortOrder("desc");
  };

  const handleToggleSort = () => {
    setIsSortOpen((prevState) => !prevState);
  };

  const handleToggleFilter = () => {
    setIsFilterOpen((prevState) => !prevState);
  };

  //not efficient if there's a large # of users
  const handleFilter = (option: string) => {
    setIsFilterOpen(false);
  
    fetch("http://localhost:3001/users")
      .then((response) => response.json())
      .then((data) => {
        let filteredData = data;
  
        if (option === "withSignature") {
          filteredData = data.filter((user: User) => user.signature !== null);
        } else if (option === "withoutSignature") {
          filteredData = data.filter((user: User) => user.signature === null);
        }
  
        setUserList(filteredData);
      })
      .catch((error) => console.error(error));
  };

  return (
    <div data-testid="toolbar-component">
      <h1 className="page-title">User Roster</h1>
      <div className="toolbar">
        <button onClick={handleOpenPopup} className="add-user-btn">
          Add New User
        </button>
        <div className="sort-container">
          <h3 className="toolbar-title">Sort By: </h3>
          <button className="sort-button" onClick={handleToggleSort}>
            Date of Birth ({sortOrder === "asc" ? "Ascending" : "Descending"})
          </button>
          {isSortOpen && (
            <div className="sort-options">
              <div className="sort-option" onClick={handleSortAscending}>
                Date of Birth (Ascending)
              </div>
              <div className="sort-option" onClick={handleSortDescending}>
                Date of Birth (Descending)
              </div>
            </div>
          )}

          <h3 className="toolbar-title">Filter By: </h3>
          <button className="sort-button" onClick={handleToggleFilter}>Filter Options</button>
          {isFilterOpen && (
            <div className="filter-options">
              <div className="filter-option" onClick={() => handleFilter("")}>
                All Users
              </div>
              <div className="filter-option" onClick={() => handleFilter("withSignature")}>
                Users With Custom Signatures
              </div>
              <div className="filter-option" onClick={() => handleFilter("withoutSignature")}>
                Users Without Custom Signatures
              </div>
            </div>
          )}
        </div>
      </div>
      {isPopupOpen && (
        <div className="popup-overlay">
          <AddUser userList={userList} onClose={handleClosePopup} onAddUser={handleAddUser} />
        </div>
      )}
    </div>
  );
};

export default Toolbar;
