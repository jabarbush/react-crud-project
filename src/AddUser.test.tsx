import { render, screen, fireEvent } from '@testing-library/react';
import User from './interfaces/User';
import Toolbar from './components/Toolbar';
import AddUser from './components/AddUser';

beforeEach(() => jest.resetAllMocks())
describe('AddUser component', () => {
    const userList: User[] = [{
        "id": 4,
        "name": "jake425525333",
        "dob": "03/03/2003",
        "phone": "4564564564",
        "email": "jake4@gmail.com",
        "signature": null
      },
      {
        "id": 5,
        "name": "ashley1234",
        "dob": "05/05/1995",
        "phone": "0980980988",
        "email": "ashley@jake.com",
        "signature": {
          "pin": "123123",
          "fontStyle": "Dancing Script"
        }
      },
      {
        "id": 6,
        "name": "123",
        "dob": "01/01/2001",
        "phone": "1231231231",
        "email": "jake@gmail.com",
        "signature": {
          "pin": "123123",
          "fontStyle": "Great Vibes"
        }
      }];

test('closes the AddUser popup when it is open and "cancel" button is clicked', () => {
    render(<Toolbar userList={[]} setUserList={jest.fn()} />);
    
    const addButton = screen.getByText('Add New User');
    fireEvent.click(addButton);
  
    const closeButton = screen.getByTestId('add-user-cancel-button');
    fireEvent.click(closeButton);
  });

  test('AddUser component matches snapshot', () => {
    const onClose = jest.fn();
    const setUserList = jest.fn();
  
    const { asFragment } = render(
      <AddUser
        userList={userList}
        onClose={onClose}
        onAddUser={setUserList}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});