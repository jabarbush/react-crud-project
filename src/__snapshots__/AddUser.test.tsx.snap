// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUser component AddUser component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="add-user-popup"
  >
    <form>
      <h2
        class="new-user-title"
      >
        Add New User
      </h2>
      <label>
        Name
      </label>
      <input
        placeholder="Name"
        required=""
        type="text"
        value=""
      />
      <label>
        Date of Birth
      </label>
      <input
        placeholder="Date of Birth"
        required=""
        type="text"
        value=""
      />
      <label>
        Phone
      </label>
      <input
        placeholder="Phone"
        required=""
        type="text"
        value=""
      />
      <label>
        Email
      </label>
      <input
        placeholder="Email"
        required=""
        type="text"
        value=""
      />
      <button
        class="cancel"
        data-testid="add-user-cancel-button"
      >
        Cancel
      </button>
      <button
        class="add-user-submit"
        type="submit"
      >
        Add User
      </button>
    </form>
  </div>
</DocumentFragment>
`;
